### PR TITLE
Correct positioning for unique token filter

### LINKED ISSUE #35411: Unique token filter results in tokens with misplaced positions
**Elasticsearch version** (`bin/elasticsearch --version`):

Version: 6.4.2, Build: oss/tar/04711c2/2018-09-26T13:34:09.098244Z, JVM: 1.8.0_181

**Plugins installed**: None

**JVM version** (`java -version`): java version "1.8.0_181"

**OS version** (`uname -a` if on a Unix-like system):

Ubuntu 16.04.5 LTS Linux 4.4.0-1069-aws #79-Ubuntu SMP Mon Sep 24 15:01:41 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
Unique token filter assigns incorrect positions to tokens that have had their first character deduplicated.

**Steps to reproduce**:

*Create index:*
```
curl -XPUT localhost:9200/test
```

*Expected:* Analysis of a text `los angeles leadership academy` using the `edge_ngram` and `unique` token filters should result in unique tokens sized at the `max_gram` (default 2) of the `edge_ngram` filter and span the positions of the 4 words.
```
curl -XPOST localhost:9200/test/_analyze -H 'Content-Type:application/json' -d '{
    "tokenizer": "standard",
    "text": "los angeles leadership academy",
    "filter": ["edge_ngram", "unique"]
}'
```

*Actual:*
The tokens `l, lo, a, an, le, ac` are produced. However, the tokens `le` and `ac` are incorrectly placed at position 1 due to the first letters in their respective words getting deduplicated by the `unique` token filter. See below:
```
{
  "tokens": [
    {
      "token": "l",
      "start_offset": 0,
      "end_offset": 3,
      "type": "<ALPHANUM>",
      "position": 0
    },
    {
      "token": "lo",
      "start_offset": 0,
      "end_offset": 3,
      "type": "<ALPHANUM>",
      "position": 0
    },
    {
      "token": "a",
      "start_offset": 4,
      "end_offset": 11,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "an",
      "start_offset": 4,
      "end_offset": 11,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "le",
      "start_offset": 12,
      "end_offset": 22,
      "type": "<ALPHANUM>",
      "position": 1
    },
    {
      "token": "ac",
      "start_offset": 23,
      "end_offset": 30,
      "type": "<ALPHANUM>",
      "position": 1
    }
  ]
}

```

A different text, `los angeles city hall` for example, would produce the tokens `l, lo, a, an, c, ci, h, ha` that correctly span positions 0 to 3.

### PR DESCRIPTION
**Pull request by benwtrent**
This is an extension of: https://github.com/elastic/elasticsearch/pull/35420

closes: https://github.com/elastic/elasticsearch/issues/35411

### CODE REVIEW

**Comment by benwtrent** [comment]
@javanna I am unsure about a new deprecated setting to revert to the old behavior. It may be overkill.

**Comment by benwtrent** [comment]
@shuaibiyy I know we are about 6 years later, but if you are still around, mind signing the CLA for this particular change? While the current evolution of the change has grown, I would still like you to get contribution credit if possible.

**Comment by shuaibiyy** [comment]
Hi @benwtrent I've signed the CLA. Thanks!

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by cbuescher — APPROVED** [review]
LGTM, I left a minor question and also quickly read over the bwc discussion and agree adding a quirks-mode setting.

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/UniqueTokenFilterTests.java [hunk 1]
```diff
@@ -32,15 +45,121 @@ protected TokenStreamComponents createComponents(String fieldName) {
...
+            @Override
+            protected TokenStreamComponents createComponents(String fieldName) {
+                Tokenizer t = new MockTokenizer(MockTokenizer.WHITESPACE, false);
+                return new TokenStreamComponents(t, new UniqueTokenFilter(t, true));
+            }
+        };
+
+        TokenStream test = analyzer.tokenStream("test", "this test with test");
+        test.reset();
+        CharTermAttribute termAttribute = test.addAttribute(CharTermAttribute.class);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/UniqueTokenFilterTests.java:75]: Yep! Added.

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/XUniqueTokenFilterTests.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class XUniqueTokenFilterTests extends ESTestCase {
+    public void testSimple() throws IOException {
+        Analyzer analyzer = new Analyzer() {
+            @Override
+            protected TokenStreamComponents createComponents(String fieldName) {
+                Tokenizer t = new MockTokenizer(MockTokenizer.WHITESPACE, false);
+                return new TokenStreamComponents(t, new UniqueTokenFilter(t));
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/XUniqueTokenFilterTests.java]: correct, fixed

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**